File: types/newsletter.ts
import type { BaseEntity } from "./common";

export type NewsletterCategory = 
  | "residential"
  | "commercial"
  | "taxation"
  | "regulation"
  | "market_trends";

export type NewsletterFrequency = "daily" | "weekly" | "monthly";

export interface Newsletter extends Omit<BaseEntity, 'organization_id'> {
  title: string;
  category: NewsletterCategory;
  frequency: NewsletterFrequency;
  description: string;
  active: boolean;
  slug?: string;
  theme?: string;
  is_active?: boolean;
  organization_id?: string;
}

export interface NewsletterWithSubscription extends Newsletter {
  is_user_subscribed?: boolean;
  subscription_id?: string;
  edition_count?: number;
  subscriber_count?: number;
  last_published_at?: string;
}

export interface NewsletterSubscription {
  id: string;
  user_id: string;
  newsletter_id: string;
  subscribed_at: string;
  unsubscribed_at?: string;
  active: boolean;
}

export interface NewsletterEdition extends BaseEntity {
  newsletter_id: string;
  title: string;
  content: string;
  published_at?: string;
  sent_at?: string;
  recipient_count?: number;
  open_rate?: number;
  click_rate?: number;
}

export interface SubscribeNewsletterRequest {
  newsletter_id: string;
}

export interface UnsubscribeNewsletterRequest {
  newsletter_id: string;
}

File: backend/schemas/newsletter.py
"""
Pydantic schemas for newsletter functionality
"""
from pydantic import BaseModel
from datetime import datetime
from typing import Optional
from uuid import UUID


class NewsletterBase(BaseModel):
    slug: str
    title: str
    description: Optional[str] = None
    theme: str
    frequency: str
    is_active: bool = True


class NewsletterCreate(NewsletterBase):
    pass


class NewsletterUpdate(BaseModel):
    slug: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    theme: Optional[str] = None
    frequency: Optional[str] = None
    is_active: Optional[bool] = None


class Newsletter(NewsletterBase):
    id: UUID
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class NewsletterEditionBase(BaseModel):
    newsletter_id: UUID
    title: str
    content: str
    published_at: datetime
    created_at: Optional[datetime] = None


class NewsletterEditionCreate(NewsletterEditionBase):
    pass


class NewsletterEditionUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    published_at: Optional[datetime] = None


class NewsletterEdition(NewsletterEditionBase):
    id: UUID
    created_at: datetime
    updated_at: datetime
    # Optional fields from database
    organization_id: Optional[UUID] = None
    sent_at: Optional[datetime] = None
    recipient_count: Optional[int] = None
    open_rate: Optional[float] = None
    click_rate: Optional[float] = None
    
    class Config:
        from_attributes = True


class NewsletterSubscriptionBase(BaseModel):
    user_id: UUID
    newsletter_id: UUID
    is_subscribed: bool = True


class NewsletterSubscriptionCreate(BaseModel):
    newsletter_id: UUID


class NewsletterSubscription(NewsletterSubscriptionBase):
    id: UUID
    subscribed_at: datetime
    unsubscribed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class NewsletterWithSubscription(Newsletter):
    is_user_subscribed: bool = False
    subscription_id: Optional[UUID] = None


class NewsletterGenerationRequest(BaseModel):
    lookback_days: Optional[int] = 30
    start_date: Optional[str] = None
    end_date: Optional[str] = None


class NewsletterGenerationResponse(BaseModel):
    success: bool
    edition_id: Optional[str] = None
    message: str
    articles_processed: Optional[int] = None
    generation_time: Optional[float] = None
